core/src/main/scala/kafka/zookeeper/ZookeeperClient.scala

```diff
@@ -279,33 +258,130 @@ trait ZNodeChildChangeHandler {
 }
 
 sealed trait AsyncRequest {
-  val path: String
-  val ctx: Any
+  type Response <: AsyncResponse
+  def path: String
+  def ctx: Option[Any]
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit)
+}
+
+sealed trait WatchableAsyncRequest extends AsyncRequest {
+
+  final protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit) =
+    send(zooKeeper, false)(processResponse)
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper, watch: Boolean)(processResponse: Response => Unit)
+}
+
+case class CreateRequest(path: String, data: Array[Byte], acl: Seq[ACL], createMode: CreateMode, ctx: Option[Any] = None) extends AsyncRequest {
+  type Response = CreateResponse
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit): Unit = {
+    zooKeeper.create(path, data, acl.asJava, createMode, new StringCallback {
+      override def processResult(rc: Int, path: String, ctx: Any, name: String): Unit =
+        processResponse(CreateResponse(Code.get(rc), path, Option(ctx), name))
+    }, ctx.orNull)
+  }
+}
+
+case class DeleteRequest(path: String, version: Int, ctx: Option[Any] = None) extends AsyncRequest {
+  type Response = DeleteResponse
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit): Unit = {
+    zooKeeper.delete(path, version, new VoidCallback {
+      override def processResult(rc: Int, path: String, ctx: Any): Unit =
+        processResponse(DeleteResponse(Code.get(rc), path, Option(ctx)))
+    }, ctx.orNull)
+  }
+}
+
+case class ExistsRequest(path: String, ctx: Option[Any] = None) extends WatchableAsyncRequest {
+  type Response = ExistsResponse
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper, watch: Boolean)(processResponse: Response => Unit): Unit = {
+    zooKeeper.exists(path, watch, new StatCallback {
+      override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) =
+        processResponse(ExistsResponse(Code.get(rc), path, Option(ctx), stat))
+    }, ctx.orNull)
+  }
+}
+
+case class GetDataRequest(path: String, shouldWatch: Option[Boolean] = None, ctx: Option[Any] = None) extends WatchableAsyncRequest {
```

Comment:
Looks like the earlier refactoring was incomplete. `shouldWatch` shouldn't exist in the constructor.